DSA_Session5/src/dictionary/DictionaryImp.java

```diff
@@ -0,0 +1,273 @@
+	}
+
+	/**
+	 * recursive function for deleting a key form BST
+	 * @param key
+	 * @param node
+	 * @return
+	 */
+	private boolean deleteRec(String key, BSTNode node) {
+		// go left if key less than current node key
+		if (key.compareTo(node.getKey()) <= -1) {
+			if (node.getLeft() != null) {
+				// if left is present call recursive
+				return deleteRec(key, node.getLeft());
+			}
+		}
+		// go right if key is greater than current node key
+		else if (key.compareTo(node.getKey()) >= 1) {
+			if (node.getRight() != null) {
+				// if right is present call recursive
+				return deleteRec(key, node.getRight());
+			}
+		}
+		// if key is equal to node key
+		else if (key.compareTo(node.getKey()) == 0) {
+			// check if it is a leaf node
+			if (node.getLeft() == null) {
+				if (node.getRight() == null) {
+					// set parent child null if it is leaf node
+					if (node.getParent().getLeft() == node) {
+						node.getParent().setLeft(null);
+					} else {
+						node.getParent().setRight(null);
+					}
+				}
+			}
+			// if left node is null
+			else if (node.getLeft() == null && node.getRight() != null) {
+				// if self is left than set parent left else set parent right
+				if (node.getParent().getLeft() == node) {
+					node.getParent().setLeft(node.getRight());
+				} else {
+					node.getParent().setRight(node.getRight());
+				}
+			}
+			// if right is null
+			else if (node.getRight() == null && node.getLeft() != null) {
+				// if self is right set parent right else set parent left
+				if (node.getParent().getRight() == node) {
+					node.getParent().setRight(node.getLeft());
+				} else {
+					node.getParent().setLeft(node.getLeft());
+				}
+			}
+
+			else {
+				BSTNode newNode = findMin(node.getRight());
+				node.setKey(newNode.getKey());
+				node.setValue(newNode.getValue());
+				return deleteRec(newNode.getKey(), newNode);
+			}
+		}
+		return true;
+	}
+
+	/**
+	 * method to find minimum value from tree
+	 * @param currentNode
+	 * @return
+	 */
+	private BSTNode findMin(BSTNode currentNode) {
+		while (currentNode.getLeft() != null) {
+			currentNode = currentNode.getLeft();
+		}
+		return currentNode;
+	}
+
+	/**
+	 * method to traverse tree
+	 * @param root
+	 * @param list
+	 * @return
+	 */
+	private List<Value> inorderRec(BSTNode root, List<Value> list) {
+		if (root != null) {
+			inorderRec(root.left, list);
+			list.add(new Value(root.getKey(), root.getValue()));
+			inorderRec(root.right, list);
+		}
+		return list;
+	}
+
+	/**
+	 * method to find a value with key 
+	 * @param key
+	 */
+	@Override
+	public String getValue(String key) {
+		if (head != null) {
+			return getValueRec(key, head);
+		}
+		return null;
+	}
+
+	/**
+	 * recursive method to find a value form tree
+	 * @param key
+	 * @param node
+	 * @return
+	 */
+	private String getValueRec(String key, BSTNode node) {
+		if(node == null) {
+			return null;
+		} else if (key.compareTo(node.getKey()) == 0) {
+			return node.getValue();
```

Comment:
keep the different name for the method in different classes, so it can be understood easily.

Reply:
please explain this.